### DISCUSSION TITLE
encodeFunctionData and similar with generics

### DISCUSSION THREAD

**Discussion by pociej** [wevm/viem#2777] [Idea / Feature Request]
H
I wanna use encodeFunctionData inside generic function 

```
function generic<T extends Abi, F extends ContractFunctionName<T>>(
  abi: T,
  functionName: F, 
  args : ContractFunctionArgs<T,AbiStateMutability,F> 
) {
  const callData = encodeFunctionData({
    abi, 
    functionName,
    args 
  })
}
``` 

And here problem comes as typescripts gives complicated error which most important part is : 

```
'{ abi: T; functionName: F; args: ContractFunctionArgs<T, AbiStateMutability, F>; }' 
is not assignable to type 
'UnionEvaluate<
   IsNarrowable<T, Abi> extends true 
     ? T["length"] extends 1 
       ? { functionName?: `0x${string}` | F | ContractFunctionName<T, AbiStateMutability> | undefined; } 
       : { functionName: `0x${string}` | F | ContractFunctionName<T, AbiStateMutability>; } 
     : { functionName?: `0x${string}` | F | ContractFunctionName<T, AbiStateMutability> | undefined; }
>'.
``` 
I checked utils implementation and seens that all the evil comes from how `UnionEvaluate` uses `Prettify`  https://github.com/wevm/viem/blob/main/src/types/utils.ts#L165 
Which is very useful in development but doesnt seem to work well with generics.  Does one of you meet and solve this problem already?

**Comment by tmm** [comment]
Couple quick options for you in this [TypeScript Playground](https://www.typescriptlang.org/play/?target=99#code/JYWwDg9gTgLgBAbwLACg5wKYDsDGEAmGAYgK64zARYAiAhjLQDSroZQ4BMADAIIBGwZmjgwAnmAxx+gliPGTpAZQYwMAWRIMBAG2Bih6MRLgBhKjCi0cMUuUpYAcrRAYDc42awWrNstfs8UADmAM5CAL5wAGZQECBwAOQAbsAYIAmoqAD0Waa0IfDAWNF+FFRwfASiqMjCUaX2cEHYbMA4ADyy6LQCmAAeqlj4IVICbuj1dlROLv2Dw6bmlta2-tPOGO09wAB8bjsAFF1w2wBcJ2PHk2uOG+fXZbcu4yfBIeee3isNVIGhW2NRsBlPR1Jptrp9CUpk8MHtZABKRDHPBYApwHC0bTaOgMOAAXkwuAIxB+NHotAOtXQNIuwHO2xOI2kLwmZJmGHu7I2TLgBSgRSCrNeoQZb15UAwtHwVG0ojgZAA1lgIAB3LAAbQAui9wgjZOFMsJmlhWjgDmxOLxAQlaGAwLEkhgEow4BqElw+i64ABGDgAZiwWv1KENKGyuQAqiFJBAnVBtBBpdFoO5JEUomxsDgMDVZA9GiazZ1hN1ehgBtgFiyrtzZhX5iNPstfDCOQDdvsjqW6QzLj2C+sXFy2xsXrQ3h8lj5Vo8-iEO66lCowVpgJDRK7B7D4cIEeckhBgPh82Smi0BebjmcgS9txzzvzBePJ3BJdLZfKlSr1dq4AAfBUhgwKIigwfA3CRakaVRdFMWxXFaAJIk8EIWd7EQqk6S3OtXBFEY9QNI10CLS8LXYbgWUSO0HTjZ1XXdT1vT9QNg1QMMgA).

### Cast in function body

Perfectly fine and natural to cast internal function values.

```ts
function generic<
  abi extends Abi,
  functionName extends ContractFunctionName<abi>,
>(
  abi: abi,
  functionName: functionName,
  args: ContractFunctionArgs<abi, AbiStateMutability, functionName>,
) {
  const callData = encodeFunctionData({
    abi: abi as Abi,
    functionName: functionName as string,
    args: args as readonly unknown[],
  })
}
```

### Use overload for type inference

Create one signature that handles type inference, while the implementation uses the non-inferred values.

```ts
function generic<
  abi extends Abi,
  functionName extends ContractFunctionName<abi>,
>(
  abi: abi,
  functionName: functionName,
  args: ContractFunctionArgs<abi, AbiStateMutability, functionName>,
): void

function generic(
  abi: Abi,
  functionName: string,
  args: readonly unknown[] | undefined,
) {
  const callData = encodeFunctionData({ abi, functionName, args })
}
```

**Reply by pociej** [reply]
Thx for reply @tmm. I see  downsides in both approach. 

> Perfectly fine and natural to cast internal function values.

 If `encodeFunnctionData` is properly typed then such  casting  could cause runtime errors. 

> Create one signature that handles type inference, while the implementation uses the non-inferred values.

In second implemenation `functionName` is of type string while should be exact name of function actually existing on contract. Look at the same [Typescript Playground](https://www.typescriptlang.org/play/?target=99#code/JYWwDg9gTgLgBAbwLACg5wKYDsDGEAmGAYgK64zARYAiAhjLQDSroZQ4BMADAIIBGwZmjgwAnmAxx+gliPGTpAZQYwMAWRIMBAG2Bih6MRLgBhKjCi0cMUuUpYAcrRAYDc42awWrNstfs8UADmAM5CAL5wAGZQECBwAOQAbsAYIAmoqAD0Waa0IfDAWNF+FFRwfASiqMjCUaX2cEHYbMA4ADyy6LQCmAAeqlj4IVICbuj1dlROLv2Dw6bmlta2-tPOGO09wAB8bjsAFF1w2wBcJ2PHk2uOG+fXZbcu4yfBIeee3isNVIGhW2NRsBlPR1Jptrp9CUpk8MHtZABKRDHPBYApwHC0bTaOgMOAAXkwuAIxB+NHotAOtXQNIuwHO2xOI2kLwmZJmGHu7I2TLgBSgRSCrNeoQZb15UAwtHwVG0ojgZAA1lgIAB3LAAbQAui9wgjZOFMsJmlhWjgDmxOLxAQlaGAwLEkhgEow4BqElw+i64ABGDgAZiwWv1KENKGyuQAqiFJBAnVBtBBpdFoO5JEUomxsDgMDVZKj0YzCRrZNSaUZOXAAEQPexVl5YO7VvhY2i4DAAeSi9eOBVBGi0wEhonOVZSGFVPeE6CKYE07zdCDTo+l+ElIRCVbg4R1xwgmjnMAXGqXFdHJCKMA4AFYAGxbnducJCLW8gswADcRorcAAogNll8GEORCdoABU5mwBZJWlWV5SXRsXHOflBW3bUdgJOAwI1LASBAPg2C1d1EOdLUvxQH9VkeEDMP-L4gJuED2grCAojpHYPzgI1a3KGVFDiDAYAAC0FAB1PQhKo+wOQOPo4HOKT1hcEIEVqVAw1QHjihNM1OmERkMAGKDmUuOpuVmQz5hGT5AMU2EAV2IRDlkM46TcLSOS5YCNjcWg3g+JYfDsv5QO2V0lBUMFB2HV0PI2eEUARc4kggYB8HIzSySaFoBXNFyBHOFlZDipC+QsQVfP8uAYJlLA5QVLBlTVTVXwAHwawgoiKDB8CEJEyz4gThLEiS7JkkqMBDdB3wxLEcQpTDsxJOzcUpJcwuhRifJFEY9Qy0MjR03KDg2qsW20Nscy7es3SrT0q2DfajXU1AgA) i just edited

**Comment by pociej** [comment]
I went through implementation and I see it shouldn't work in case I described in initial post as generally there could `Abi` without function. Anyways look at this, input types should work and it still produces mentioned TS error. 

``` 
export async function generic<
  T extends readonly [AbiFunction, ...Abi[number][]],
  F extends ContractFunctionName<T>,
>(
  abi: T,
  functionName: F,
  args: ContractFunctionArgs<T, AbiStateMutability, F>,
) {
  const calldata = encodeFunctionData({
    abi: abi,
    functionName: functionName,
    args: args,
  });
}
```
With small refactor in viem it should work, however it would most probably need us to resign from using `Prettify` under the hood. 
I could create draft PR to show idea if you are open to consider such changes.
